bob-access/src/credentials.rs

```diff
@@ -59,64 +73,186 @@ impl CredentialsKind {
+    pub fn with_address(mut self, address: Option<SocketAddr>) -> Self {
         self.address = address;
         self
     }
 
-    pub fn with_nodename(&mut self, node_name: impl Into<String>) -> &mut Self {
+    pub fn with_nodename(mut self, node_name: impl Into<String>) -> Self {
         self.kind = Some(CredentialsKind::InterNode(node_name.into()));
         self
     }
 
-    pub fn build(&self) -> Credentials {
-        Credentials {
+    pub fn build(self) -> RequestCredentials {
+        RequestCredentials {
+            address: self.address,
+            kind: self.kind,
+        }
+    }
+}
+
+const RESOLVE_THRESHOLD: u32 = 3;
+
+#[derive(Debug, Clone)]
+enum ResolveState {
+    Waiting(u32),
+    Pending,
+}
+
+impl ResolveState {
+    fn update(&mut self, unresolved: bool) -> bool {
+        match self {
+            ResolveState::Waiting(n) => {
+                if unresolved {
+                    *n += 1;
+                    *n >= RESOLVE_THRESHOLD
+                } else {
+                    *n = 0;
+                    false
+                }                
+            },
+            ResolveState::Pending => false,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DeclaredCredentials {
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+    kind: CredentialsKind,
+}
+
+impl DeclaredCredentials {
+    pub fn internode_builder(node_name: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::InterNode(node_name.into()))
+    }
+
+    pub fn token_builder(token: impl Into<String>) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Token(token.into()))
+    }
+
+    pub fn userpass_builder(username: impl Into<String>, password: impl Into<String>,) -> DeclaredCredentialsBuilder {
+        DeclaredCredentialsBuilder::new(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        })
+    }
+
+    pub fn ip(&self) -> &Vec<SocketAddr> {
+        &self.address
+    }
+
+    pub fn replace_addresses(&mut self, addresses: Vec<SocketAddr>) {
+        self.address = addresses;
+    }
+
+    pub fn hostname(&self) -> &Option<String> {
+        &self.hostname
+    }
+
+    pub fn validate_internode(&self) -> bool {
+        (!self.address.is_empty() || self.hostname.is_some()) &&
+        self.kind.is_internode()
+    }
+
+    pub fn kind(&self) -> &CredentialsKind {
+        &self.kind
+    }
+}
+
+#[derive(Debug)]
+pub struct DeclaredCredentialsBuilder {
+    kind: CredentialsKind,
+    address: Vec<SocketAddr>,
+    hostname: Option<String>,
+}
+
+impl DeclaredCredentialsBuilder {
+    fn new(kind: CredentialsKind) -> Self {
+        Self {
+            kind,
+            address: Vec::new(),
+            hostname: None,
+        }
+    }
+
+    pub fn with_address(mut self, address: SocketAddr) -> Self {
+        self.address.push(address);
```

Comment:
This is not what I meant. I meant, that `with_address` would init a vector with single value, not update existing one by pushing a new item